table/temptable/interceptor.go

```diff
@@ -0,0 +1,257 @@
+		return i.iterTable(tblID, snap, k, upperBound)
+	}
+
+	return createUnionIter(i.sessionData, snap, k, upperBound, false)
+}
+
+// OnIterReverse intercepts IterReverse operation for Snapshot
+func (i *TemporaryTableSnapshotInterceptor) OnIterReverse(snap kv.Snapshot, k kv.Key) (kv.Iterator, error) {
+	if safeWithSnapshotScan(nil, k) {
+		// scan range has no intersect with table data
+		return snap.IterReverse(k)
+	}
+
+	// because lower bound is nil, so the range cannot be located in one table
+	return createUnionIter(i.sessionData, snap, nil, k, true)
+}
+
+func (i *TemporaryTableSnapshotInterceptor) iterTable(tblID int64, snap kv.Snapshot, k, upperBound kv.Key) (kv.Iterator, error) {
+	tblInfo, ok := i.temporaryTableInfoByID(tblID)
+	if !ok {
+		return snap.Iter(k, upperBound)
+	}
+
+	if tblInfo.TempTableType == model.TempTableGlobal || i.sessionData == nil {
+		return &kv.EmptyIterator{}, nil
+	}
+
+	// still need union iter to filter out empty value in session data
+	return createUnionIter(i.sessionData, nil, k, upperBound, false)
+}
+
+func (i *TemporaryTableSnapshotInterceptor) temporaryTableInfoByID(tblID int64) (*model.TableInfo, bool) {
+	if tbl, ok := i.is.TableByID(tblID); ok {
+		tblInfo := tbl.Meta()
+		if tblInfo.TempTableType != model.TempTableNone {
+			return tblInfo, true
+		}
+	}
+
+	return nil, false
+}
+
+func createUnionIter(sessionData kv.Retriever, snap kv.Snapshot, k, upperBound kv.Key, reverse bool) (iter kv.Iterator, err error) {
+	if reverse && k != nil {
+		return nil, errors.New("k should be nil for iter reverse")
+	}
+
+	var snapIter kv.Iterator
+	if snap == nil {
+		snapIter = &kv.EmptyIterator{}
+	} else {
+		if reverse {
+			snapIter, err = snap.IterReverse(upperBound)
+		} else {
+			snapIter, err = snap.Iter(k, upperBound)
+		}
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	if sessionData == nil {
+		return snapIter, nil
+	}
+
+	var sessionIter kv.Iterator
+	if reverse {
+		sessionIter, err = sessionData.IterReverse(upperBound)
+	} else {
+		sessionIter, err = sessionData.Iter(k, upperBound)
+	}
+
+	if err != nil {
+		snapIter.Close()
+		return nil, err
+	}
+
+	iter, err = txn.NewUnionIter(sessionIter, snapIter, reverse)
+	if err != nil {
+		snapIter.Close()
+		sessionIter.Close()
+	}
+	return iter, err
+}
+
+func getRangeAccessedTableID(startKey, endKey kv.Key) (int64, bool) {
+	tblID, ok := getKeyAccessedTableID(startKey)
+	if !ok {
+		return 0, false
+	}
+
+	tblStart := tablecodec.EncodeTablePrefix(tblID)
+	tblEnd := tablecodec.EncodeTablePrefix(tblID + 1)
+	if bytes.HasPrefix(endKey, tblStart) || bytes.Equal(endKey, tblEnd) {
+		return tblID, true
+	}
+
+	return 0, false
+}
+
+func getKeyAccessedTableID(k kv.Key) (int64, bool) {
+	if bytes.HasPrefix(k, tablecodec.TablePrefix()) && len(k) >= tableWithIDPrefixLen {
+		if tbID := tablecodec.DecodeTableID(k); tbID > 0 && tbID != math.MaxInt64 {
+			return tbID, true
+		}
+	}
+
+	return 0, false
+}
+
+func safeWithSnapshotScan(k, upperBound kv.Key) bool {
```

Comment:
Is `notTableRange` a better name?

Reply:
OK. It's too hard for me to choose with name is better...